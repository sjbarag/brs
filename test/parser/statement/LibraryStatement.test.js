const brs = require("brs");
const { Lexeme } = brs.lexer;
const { BrsString, Int32 } = brs.types;

const { token, identifier, EOF } = require("../ParserTests");

describe("parser library statements", () => {
    let parser;

    beforeEach(() => {
        parser = new brs.parser.Parser();
    });

    it('supports library statements at top of file', () => {
        const { tokens } = brs.lexer.Lexer.scan(`
            Library "v30/bslCore.brs"
            sub main()
            end sub
        `);
        const { statements, errors } = parser.parse(tokens);
        expect(errors.length).toEqual(0);
        expect({ errors, statements }).toMatchSnapshot();
    });

    it('supports multiple library statements separated by colon', () => {
        const { tokens } = brs.lexer.Lexer.scan(`
            Library "v30/bslCore.brs" : Library "v30/bslCore.brs"
            sub main()
            end sub
        `);
        const { statements, errors } = parser.parse(tokens);
        expect(errors.length).toEqual(0);
        expect({ errors, statements }).toMatchSnapshot();
    });

    it('adds error for library statements NOT at top of file', () => {
        const { tokens } = brs.lexer.Lexer.scan(`
            sub main()
            end sub
            Library "v30/bslCore.brs"
        `);
        const { statements, errors } = parser.parse(tokens);
        expect(errors.length).toBeGreaterThan(0);
        expect({ errors, statements }).toMatchSnapshot();
    });

    it('adds error for library statements inside of function body', () => {
        const { tokens } = brs.lexer.Lexer.scan(`
            sub main()
                Library "v30/bslCore.brs"
            end sub
        `);
        const { statements, errors } = parser.parse(tokens);
        expect(errors.length).toEqual(1);
        expect({ errors, statements }).toMatchSnapshot();
    });

    it('still parses entire file after invalid library statement', () => {
        const { tokens } = brs.lexer.Lexer.scan(`
            library cat dog mouse
            sub main()
            end sub
        `);
        const { statements, errors } = parser.parse(tokens);
        expect(errors.length).toBeGreaterThan(0);
        expect({ errors, statements }).toMatchSnapshot();
    });

    it('does not prevent usage of `library` as varible name', () => {
        const { tokens } = brs.lexer.Lexer.scan(`
            sub main()
                library = "Gotham City Library"
            end sub
        `);
        const { statements, errors } = parser.parse(tokens);
        //make sure the assignment is present in the function body
        let assignment = statements[0].func.body.statements[0];
        expect(assignment).toBeInstanceOf(brs.parser.Stmt.Assignment)
        expect(assignment.name.text).toEqual('library');
        expect({ errors, statements }).toMatchSnapshot();
    });

    it('does not prevent usage of `library` as object property name', () => {
        const { tokens } = brs.lexer.Lexer.scan(`
            sub main()
                buildings = {
                    library: "Gotham City Library"
                }
            end sub
        `);
        const { statements, errors } = parser.parse(tokens);
        //make sure the assignment is present in the function body
        expect(statements[0].func.body.statements[0].value.elements[0].name.value).toEqual('library');
        expect({ errors, statements }).toMatchSnapshot();
    });

    it('parses rest of file with ONLY the library keyword present at root level', () => {
        const { tokens } = brs.lexer.Lexer.scan(`
            library
            sub main()
                library = "Your Library"
            end sub
        `);
        const { statements, errors } = parser.parse(tokens);
        expect(errors.length).toEqual(1);
        //function statement should still exist
        expect(statements[statements.length - 1]).toBeInstanceOf(brs.parser.Stmt.Function);
        expect({ errors, statements }).toMatchSnapshot();
    });
});
