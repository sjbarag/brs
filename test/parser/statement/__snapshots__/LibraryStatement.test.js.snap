// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parser library statements adds error for library statements NOT at top of file 1`] = `
Object {
  "errors": Array [
    [Error: Library statements may only appear at the top of a file],
  ],
  "statements": Array [
    Function {
      "func": Function {
        "body": Block {
          "location": Object {
            "end": Object {
              "column": 12,
              "line": 3,
            },
            "file": "",
            "start": Object {
              "column": 22,
              "line": 2,
            },
          },
          "statements": Array [],
          "type": "Block",
        },
        "end": Object {
          "isReserved": false,
          "kind": "EndSub",
          "literal": undefined,
          "location": Object {
            "end": Object {
              "column": 19,
              "line": 3,
            },
            "file": "",
            "start": Object {
              "column": 12,
              "line": 3,
            },
          },
          "text": "end sub",
        },
        "keyword": Object {
          "isReserved": true,
          "kind": "Sub",
          "literal": undefined,
          "location": Object {
            "end": Object {
              "column": 15,
              "line": 2,
            },
            "file": "",
            "start": Object {
              "column": 12,
              "line": 2,
            },
          },
          "text": "sub",
        },
        "parameters": Array [],
        "returns": 11,
        "type": "Function",
      },
      "name": Object {
        "isReserved": false,
        "kind": "Identifier",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": 20,
            "line": 2,
          },
          "file": "",
          "start": Object {
            "column": 16,
            "line": 2,
          },
        },
        "text": "main",
      },
      "type": "Function",
    },
    Library {
      "tokens": Object {
        "filePath": Object {
          "isReserved": false,
          "kind": "String",
          "literal": BrsString {
            "kind": 3,
            "value": "v30/bslCore.brs",
          },
          "location": Object {
            "end": Object {
              "column": 37,
              "line": 4,
            },
            "file": "",
            "start": Object {
              "column": 20,
              "line": 4,
            },
          },
          "text": "\\"v30/bslCore.brs\\"",
        },
        "library": Object {
          "isReserved": false,
          "kind": "Identifier",
          "literal": undefined,
          "location": Object {
            "end": Object {
              "column": 19,
              "line": 4,
            },
            "file": "",
            "start": Object {
              "column": 12,
              "line": 4,
            },
          },
          "text": "Library",
        },
      },
      "type": "Library",
    },
  ],
}
`;

exports[`parser library statements adds error for library statements inside of function body 1`] = `
Object {
  "errors": Array [
    [Error: Library statements may only appear at the top of a file],
  ],
  "statements": Array [
    Function {
      "func": Function {
        "body": Block {
          "location": Object {
            "end": Object {
              "column": 12,
              "line": 4,
            },
            "file": "",
            "start": Object {
              "column": 22,
              "line": 2,
            },
          },
          "statements": Array [
            Library {
              "tokens": Object {
                "filePath": Object {
                  "isReserved": false,
                  "kind": "String",
                  "literal": BrsString {
                    "kind": 3,
                    "value": "v30/bslCore.brs",
                  },
                  "location": Object {
                    "end": Object {
                      "column": 41,
                      "line": 3,
                    },
                    "file": "",
                    "start": Object {
                      "column": 24,
                      "line": 3,
                    },
                  },
                  "text": "\\"v30/bslCore.brs\\"",
                },
                "library": Object {
                  "isReserved": false,
                  "kind": "Identifier",
                  "literal": undefined,
                  "location": Object {
                    "end": Object {
                      "column": 23,
                      "line": 3,
                    },
                    "file": "",
                    "start": Object {
                      "column": 16,
                      "line": 3,
                    },
                  },
                  "text": "Library",
                },
              },
              "type": "Library",
            },
          ],
          "type": "Block",
        },
        "end": Object {
          "isReserved": false,
          "kind": "EndSub",
          "literal": undefined,
          "location": Object {
            "end": Object {
              "column": 19,
              "line": 4,
            },
            "file": "",
            "start": Object {
              "column": 12,
              "line": 4,
            },
          },
          "text": "end sub",
        },
        "keyword": Object {
          "isReserved": true,
          "kind": "Sub",
          "literal": undefined,
          "location": Object {
            "end": Object {
              "column": 15,
              "line": 2,
            },
            "file": "",
            "start": Object {
              "column": 12,
              "line": 2,
            },
          },
          "text": "sub",
        },
        "parameters": Array [],
        "returns": 11,
        "type": "Function",
      },
      "name": Object {
        "isReserved": false,
        "kind": "Identifier",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": 20,
            "line": 2,
          },
          "file": "",
          "start": Object {
            "column": 16,
            "line": 2,
          },
        },
        "text": "main",
      },
      "type": "Function",
    },
  ],
}
`;

exports[`parser library statements does not prevent usage of \`library\` as object property name 1`] = `
Object {
  "errors": Array [],
  "statements": Array [
    Function {
      "func": Function {
        "body": Block {
          "location": Object {
            "end": Object {
              "column": 12,
              "line": 6,
            },
            "file": "",
            "start": Object {
              "column": 22,
              "line": 2,
            },
          },
          "statements": Array [
            Assignment {
              "name": Object {
                "isReserved": false,
                "kind": "Identifier",
                "literal": undefined,
                "location": Object {
                  "end": Object {
                    "column": 25,
                    "line": 3,
                  },
                  "file": "",
                  "start": Object {
                    "column": 16,
                    "line": 3,
                  },
                },
                "text": "buildings",
              },
              "tokens": Object {
                "equals": Object {
                  "isReserved": false,
                  "kind": "Equal",
                  "literal": undefined,
                  "location": Object {
                    "end": Object {
                      "column": 27,
                      "line": 3,
                    },
                    "file": "",
                    "start": Object {
                      "column": 26,
                      "line": 3,
                    },
                  },
                  "text": "=",
                },
              },
              "type": "Assignment",
              "value": AALiteral {
                "close": Object {
                  "isReserved": false,
                  "kind": "RightBrace",
                  "literal": undefined,
                  "location": Object {
                    "end": Object {
                      "column": 17,
                      "line": 5,
                    },
                    "file": "",
                    "start": Object {
                      "column": 16,
                      "line": 5,
                    },
                  },
                  "text": "}",
                },
                "elements": Array [
                  Object {
                    "name": BrsString {
                      "kind": 3,
                      "value": "library",
                    },
                    "value": Literal {
                      "_location": Object {
                        "end": Object {
                          "column": 50,
                          "line": 4,
                        },
                        "file": "",
                        "start": Object {
                          "column": 29,
                          "line": 4,
                        },
                      },
                      "type": "Literal",
                      "value": BrsString {
                        "kind": 3,
                        "value": "Gotham City Library",
                      },
                    },
                  },
                ],
                "open": Object {
                  "isReserved": false,
                  "kind": "LeftBrace",
                  "literal": undefined,
                  "location": Object {
                    "end": Object {
                      "column": 29,
                      "line": 3,
                    },
                    "file": "",
                    "start": Object {
                      "column": 28,
                      "line": 3,
                    },
                  },
                  "text": "{",
                },
                "type": "AALiteral",
              },
            },
          ],
          "type": "Block",
        },
        "end": Object {
          "isReserved": false,
          "kind": "EndSub",
          "literal": undefined,
          "location": Object {
            "end": Object {
              "column": 19,
              "line": 6,
            },
            "file": "",
            "start": Object {
              "column": 12,
              "line": 6,
            },
          },
          "text": "end sub",
        },
        "keyword": Object {
          "isReserved": true,
          "kind": "Sub",
          "literal": undefined,
          "location": Object {
            "end": Object {
              "column": 15,
              "line": 2,
            },
            "file": "",
            "start": Object {
              "column": 12,
              "line": 2,
            },
          },
          "text": "sub",
        },
        "parameters": Array [],
        "returns": 11,
        "type": "Function",
      },
      "name": Object {
        "isReserved": false,
        "kind": "Identifier",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": 20,
            "line": 2,
          },
          "file": "",
          "start": Object {
            "column": 16,
            "line": 2,
          },
        },
        "text": "main",
      },
      "type": "Function",
    },
  ],
}
`;

exports[`parser library statements does not prevent usage of \`library\` as varible name 1`] = `
Object {
  "errors": Array [],
  "statements": Array [
    Function {
      "func": Function {
        "body": Block {
          "location": Object {
            "end": Object {
              "column": 12,
              "line": 4,
            },
            "file": "",
            "start": Object {
              "column": 22,
              "line": 2,
            },
          },
          "statements": Array [
            Assignment {
              "name": Object {
                "isReserved": false,
                "kind": "Identifier",
                "literal": undefined,
                "location": Object {
                  "end": Object {
                    "column": 23,
                    "line": 3,
                  },
                  "file": "",
                  "start": Object {
                    "column": 16,
                    "line": 3,
                  },
                },
                "text": "library",
              },
              "tokens": Object {
                "equals": Object {
                  "isReserved": false,
                  "kind": "Equal",
                  "literal": undefined,
                  "location": Object {
                    "end": Object {
                      "column": 25,
                      "line": 3,
                    },
                    "file": "",
                    "start": Object {
                      "column": 24,
                      "line": 3,
                    },
                  },
                  "text": "=",
                },
              },
              "type": "Assignment",
              "value": Literal {
                "_location": Object {
                  "end": Object {
                    "column": 47,
                    "line": 3,
                  },
                  "file": "",
                  "start": Object {
                    "column": 26,
                    "line": 3,
                  },
                },
                "type": "Literal",
                "value": BrsString {
                  "kind": 3,
                  "value": "Gotham City Library",
                },
              },
            },
          ],
          "type": "Block",
        },
        "end": Object {
          "isReserved": false,
          "kind": "EndSub",
          "literal": undefined,
          "location": Object {
            "end": Object {
              "column": 19,
              "line": 4,
            },
            "file": "",
            "start": Object {
              "column": 12,
              "line": 4,
            },
          },
          "text": "end sub",
        },
        "keyword": Object {
          "isReserved": true,
          "kind": "Sub",
          "literal": undefined,
          "location": Object {
            "end": Object {
              "column": 15,
              "line": 2,
            },
            "file": "",
            "start": Object {
              "column": 12,
              "line": 2,
            },
          },
          "text": "sub",
        },
        "parameters": Array [],
        "returns": 11,
        "type": "Function",
      },
      "name": Object {
        "isReserved": false,
        "kind": "Identifier",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": 20,
            "line": 2,
          },
          "file": "",
          "start": Object {
            "column": 16,
            "line": 2,
          },
        },
        "text": "main",
      },
      "type": "Function",
    },
  ],
}
`;

exports[`parser library statements parses rest of file with ONLY the library keyword present at root level 1`] = `
Object {
  "errors": Array [
    [Error: Missing string literal after library keyword],
  ],
  "statements": Array [
    Library {
      "tokens": Object {
        "filePath": undefined,
        "library": Object {
          "isReserved": false,
          "kind": "Identifier",
          "literal": undefined,
          "location": Object {
            "end": Object {
              "column": 19,
              "line": 2,
            },
            "file": "",
            "start": Object {
              "column": 12,
              "line": 2,
            },
          },
          "text": "library",
        },
      },
      "type": "Library",
    },
    Function {
      "func": Function {
        "body": Block {
          "location": Object {
            "end": Object {
              "column": 12,
              "line": 5,
            },
            "file": "",
            "start": Object {
              "column": 22,
              "line": 3,
            },
          },
          "statements": Array [
            Assignment {
              "name": Object {
                "isReserved": false,
                "kind": "Identifier",
                "literal": undefined,
                "location": Object {
                  "end": Object {
                    "column": 23,
                    "line": 4,
                  },
                  "file": "",
                  "start": Object {
                    "column": 16,
                    "line": 4,
                  },
                },
                "text": "library",
              },
              "tokens": Object {
                "equals": Object {
                  "isReserved": false,
                  "kind": "Equal",
                  "literal": undefined,
                  "location": Object {
                    "end": Object {
                      "column": 25,
                      "line": 4,
                    },
                    "file": "",
                    "start": Object {
                      "column": 24,
                      "line": 4,
                    },
                  },
                  "text": "=",
                },
              },
              "type": "Assignment",
              "value": Literal {
                "_location": Object {
                  "end": Object {
                    "column": 40,
                    "line": 4,
                  },
                  "file": "",
                  "start": Object {
                    "column": 26,
                    "line": 4,
                  },
                },
                "type": "Literal",
                "value": BrsString {
                  "kind": 3,
                  "value": "Your Library",
                },
              },
            },
          ],
          "type": "Block",
        },
        "end": Object {
          "isReserved": false,
          "kind": "EndSub",
          "literal": undefined,
          "location": Object {
            "end": Object {
              "column": 19,
              "line": 5,
            },
            "file": "",
            "start": Object {
              "column": 12,
              "line": 5,
            },
          },
          "text": "end sub",
        },
        "keyword": Object {
          "isReserved": true,
          "kind": "Sub",
          "literal": undefined,
          "location": Object {
            "end": Object {
              "column": 15,
              "line": 3,
            },
            "file": "",
            "start": Object {
              "column": 12,
              "line": 3,
            },
          },
          "text": "sub",
        },
        "parameters": Array [],
        "returns": 11,
        "type": "Function",
      },
      "name": Object {
        "isReserved": false,
        "kind": "Identifier",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": 20,
            "line": 3,
          },
          "file": "",
          "start": Object {
            "column": 16,
            "line": 3,
          },
        },
        "text": "main",
      },
      "type": "Function",
    },
  ],
}
`;

exports[`parser library statements still parses entire file after invalid library statement 1`] = `
Object {
  "errors": Array [
    [Error: Found unexpected token 'cat' after library statement],
    [Error: Found unexpected token 'dog' after library statement],
    [Error: Found unexpected token 'mouse' after library statement],
  ],
  "statements": Array [
    Library {
      "tokens": Object {
        "filePath": undefined,
        "library": Object {
          "isReserved": false,
          "kind": "Identifier",
          "literal": undefined,
          "location": Object {
            "end": Object {
              "column": 19,
              "line": 2,
            },
            "file": "",
            "start": Object {
              "column": 12,
              "line": 2,
            },
          },
          "text": "library",
        },
      },
      "type": "Library",
    },
    Function {
      "func": Function {
        "body": Block {
          "location": Object {
            "end": Object {
              "column": 12,
              "line": 4,
            },
            "file": "",
            "start": Object {
              "column": 22,
              "line": 3,
            },
          },
          "statements": Array [],
          "type": "Block",
        },
        "end": Object {
          "isReserved": false,
          "kind": "EndSub",
          "literal": undefined,
          "location": Object {
            "end": Object {
              "column": 19,
              "line": 4,
            },
            "file": "",
            "start": Object {
              "column": 12,
              "line": 4,
            },
          },
          "text": "end sub",
        },
        "keyword": Object {
          "isReserved": true,
          "kind": "Sub",
          "literal": undefined,
          "location": Object {
            "end": Object {
              "column": 15,
              "line": 3,
            },
            "file": "",
            "start": Object {
              "column": 12,
              "line": 3,
            },
          },
          "text": "sub",
        },
        "parameters": Array [],
        "returns": 11,
        "type": "Function",
      },
      "name": Object {
        "isReserved": false,
        "kind": "Identifier",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": 20,
            "line": 3,
          },
          "file": "",
          "start": Object {
            "column": 16,
            "line": 3,
          },
        },
        "text": "main",
      },
      "type": "Function",
    },
  ],
}
`;

exports[`parser library statements supports library statements at top of file 1`] = `
Object {
  "errors": Array [],
  "statements": Array [
    Library {
      "tokens": Object {
        "filePath": Object {
          "isReserved": false,
          "kind": "String",
          "literal": BrsString {
            "kind": 3,
            "value": "v30/bslCore.brs",
          },
          "location": Object {
            "end": Object {
              "column": 37,
              "line": 2,
            },
            "file": "",
            "start": Object {
              "column": 20,
              "line": 2,
            },
          },
          "text": "\\"v30/bslCore.brs\\"",
        },
        "library": Object {
          "isReserved": false,
          "kind": "Identifier",
          "literal": undefined,
          "location": Object {
            "end": Object {
              "column": 19,
              "line": 2,
            },
            "file": "",
            "start": Object {
              "column": 12,
              "line": 2,
            },
          },
          "text": "Library",
        },
      },
      "type": "Library",
    },
    Function {
      "func": Function {
        "body": Block {
          "location": Object {
            "end": Object {
              "column": 12,
              "line": 4,
            },
            "file": "",
            "start": Object {
              "column": 22,
              "line": 3,
            },
          },
          "statements": Array [],
          "type": "Block",
        },
        "end": Object {
          "isReserved": false,
          "kind": "EndSub",
          "literal": undefined,
          "location": Object {
            "end": Object {
              "column": 19,
              "line": 4,
            },
            "file": "",
            "start": Object {
              "column": 12,
              "line": 4,
            },
          },
          "text": "end sub",
        },
        "keyword": Object {
          "isReserved": true,
          "kind": "Sub",
          "literal": undefined,
          "location": Object {
            "end": Object {
              "column": 15,
              "line": 3,
            },
            "file": "",
            "start": Object {
              "column": 12,
              "line": 3,
            },
          },
          "text": "sub",
        },
        "parameters": Array [],
        "returns": 11,
        "type": "Function",
      },
      "name": Object {
        "isReserved": false,
        "kind": "Identifier",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": 20,
            "line": 3,
          },
          "file": "",
          "start": Object {
            "column": 16,
            "line": 3,
          },
        },
        "text": "main",
      },
      "type": "Function",
    },
  ],
}
`;

exports[`parser library statements supports multiple library statements separated by colon 1`] = `
Object {
  "errors": Array [],
  "statements": Array [
    Library {
      "tokens": Object {
        "filePath": Object {
          "isReserved": false,
          "kind": "String",
          "literal": BrsString {
            "kind": 3,
            "value": "v30/bslCore.brs",
          },
          "location": Object {
            "end": Object {
              "column": 37,
              "line": 2,
            },
            "file": "",
            "start": Object {
              "column": 20,
              "line": 2,
            },
          },
          "text": "\\"v30/bslCore.brs\\"",
        },
        "library": Object {
          "isReserved": false,
          "kind": "Identifier",
          "literal": undefined,
          "location": Object {
            "end": Object {
              "column": 19,
              "line": 2,
            },
            "file": "",
            "start": Object {
              "column": 12,
              "line": 2,
            },
          },
          "text": "Library",
        },
      },
      "type": "Library",
    },
    Library {
      "tokens": Object {
        "filePath": Object {
          "isReserved": false,
          "kind": "String",
          "literal": BrsString {
            "kind": 3,
            "value": "v30/bslCore.brs",
          },
          "location": Object {
            "end": Object {
              "column": 65,
              "line": 2,
            },
            "file": "",
            "start": Object {
              "column": 48,
              "line": 2,
            },
          },
          "text": "\\"v30/bslCore.brs\\"",
        },
        "library": Object {
          "isReserved": false,
          "kind": "Identifier",
          "literal": undefined,
          "location": Object {
            "end": Object {
              "column": 47,
              "line": 2,
            },
            "file": "",
            "start": Object {
              "column": 40,
              "line": 2,
            },
          },
          "text": "Library",
        },
      },
      "type": "Library",
    },
    Function {
      "func": Function {
        "body": Block {
          "location": Object {
            "end": Object {
              "column": 12,
              "line": 4,
            },
            "file": "",
            "start": Object {
              "column": 22,
              "line": 3,
            },
          },
          "statements": Array [],
          "type": "Block",
        },
        "end": Object {
          "isReserved": false,
          "kind": "EndSub",
          "literal": undefined,
          "location": Object {
            "end": Object {
              "column": 19,
              "line": 4,
            },
            "file": "",
            "start": Object {
              "column": 12,
              "line": 4,
            },
          },
          "text": "end sub",
        },
        "keyword": Object {
          "isReserved": true,
          "kind": "Sub",
          "literal": undefined,
          "location": Object {
            "end": Object {
              "column": 15,
              "line": 3,
            },
            "file": "",
            "start": Object {
              "column": 12,
              "line": 3,
            },
          },
          "text": "sub",
        },
        "parameters": Array [],
        "returns": 11,
        "type": "Function",
      },
      "name": Object {
        "isReserved": false,
        "kind": "Identifier",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": 20,
            "line": 3,
          },
          "file": "",
          "start": Object {
            "column": 16,
            "line": 3,
          },
        },
        "text": "main",
      },
      "type": "Function",
    },
  ],
}
`;
